bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java

```diff
@@ -12,156 +12,417 @@
+                LOGGER.error("Not supported optional HomeKit characteristic. Service type {}, characteristic type {}",
+                        service.getType(), type, e);
+            }
+        });
+    }
+
+    /**
+     * collect optional HomeKit characteristics for a OH item.
+     * 
+     * @param taggedItem main OH item
+     * @param metadataRegistry OH metadata registry
+     * @return a map with characteristics and corresponding OH items
+     */
+    private static Map<HomekitCharacteristicType, GenericItem> getOptionalCharacteristics(HomekitTaggedItem taggedItem,
+            MetadataRegistry metadataRegistry) {
+        LOGGER.debug("get optional characteristics for item {}: isGroup? {}, isMember? {}", taggedItem.getName(),
+                taggedItem.isGroup(), taggedItem.isMemberOfAccessoryGroup());
+        Map<HomekitCharacteristicType, GenericItem> characteristicItems = new HashMap<>();
         if (taggedItem.isGroup()) {
             GroupItem groupItem = (GroupItem) taggedItem.getItem();
-            Map<HomekitCharacteristicType, Item> characteristicItems = new HashMap<>();
             groupItem.getMembers().forEach(item -> {
-                HomekitCharacteristicType type = HomekitCharacteristicType.fromItem(item);
-                if (type != null) {
-                    if (characteristicItems.containsKey(type)) {
-                        LOGGER.warn("incorrect configuration for {} detected: {} and {} are tagged as {}, skipping {}",
-                                taggedItem.getItem().getUID(), characteristicItems.get(type).getUID(), item.getUID(),
-                                type, item.getUID());
-                    } else {
-                        characteristicItems.put(type, item);
-                    }
-                }
+                getAccessoryTypes(item, metadataRegistry).stream().filter(c -> !isRootAccessory(c)).filter(
+                        c -> !isMandatoryCharacteristic(taggedItem.getAccessoryType(), legacyCheck(c.getValue())))
+                        .forEach(characteristic -> characteristicItems.put(legacyCheck(characteristic.getValue()),
+                                (GenericItem) item));
             });
-            return Collections.unmodifiableMap(characteristicItems);
         } else {
-            // do nothing; only accessory groups have characteristic items
-            return Collections.emptyMap();
+            getAccessoryTypes(taggedItem.getItem(), metadataRegistry).stream().filter(c -> !isRootAccessory(c))
+                    .filter(c -> !isMandatoryCharacteristic(taggedItem.getAccessoryType(), legacyCheck(c.getValue())))
+                    .forEach(characteristic -> characteristicItems.put(legacyCheck(characteristic.getValue()),
+                            (GenericItem) taggedItem.getItem()));
+
+            if (implicitOptionalCharacteristics.get(taggedItem.getAccessoryType()) != null) {
```

Comment:
You should cache `implicitOptionalCharacteristics.get(taggedItem.getAccessoryType())` in a local variable for reuse.